File: Koans/VBNet/AboutArrays.vb
Imports Xunit

Public Class AboutArrays
    Inherits Koan

    <Koan(1)> _
    Public Sub CreatingArrays()
        'Arrays can be created by adding () to the variable name
        Dim empty_array = New Object() {}

        Assert.Equal(FILL_ME_IN, empty_array.Length)
    End Sub

    <Koan(2)> _
    Public Sub CreatingArraysWithASize()
        ' You can also make it non-empty by adding the maximum index
        Dim array(4) As Object
        Assert.Equal(FILL_ME_IN, array.Length)
        ' Note that arrays are 0-based:
        array(0) = "abc"
        array(4) = "xyz"
        Assert.Equal(FILL_ME_IN, array(0))
        Assert.Equal(FILL_ME_IN, array(4))
        'This doesn't work:
        'Assert.Equal(FILL_ME_IN, array(-1))
    End Sub

    <Koan(3)> _
    Public Sub ArrayLiterals()
        'You don't have to specify a type if the arguments can be inferred
        Dim array = {42, 3, 2, 4}
        Assert.Equal(New Integer() {42}, array)
        Assert.True(array.IsFixedSize)
        '...it means we can't do this: array(4) = 13
        Assert.Throws(GetType(FillMeIn), Sub() array(4) = 13)
        'This is because the array is fixed at length 4.
    End Sub

    <Koan(4)> _
    Public Sub SlicingArrays()
        Dim array = New String() {"peanut", "butter", "and", "jelly"}
        Assert.Equal(New String() {FILL_ME_IN, FILL_ME_IN}, array.Take(2).ToArray())
        Assert.Equal(New String() {FILL_ME_IN, FILL_ME_IN}, array.Skip(1).Take(2).ToArray())
    End Sub

    <Koan(5)> _
    Public Sub ResizingWithReDim()
        ' You can make an array bigger or smaller
        Dim array = {42, 3, 2, 4}
        ReDim array(6)
        Assert.Equal(FILL_ME_IN, array.Length)
        ' but the values are not kept
        Assert.Equal(FILL_ME_IN, array(0))
    End Sub

    <Koan(6)> _
    Public Sub ResizingWithReDimAndPreserve()
        ' Adding Preserve keeps the values
        Dim array = {42, 3, 2, 4}
        ReDim Preserve array(6)
        Assert.Equal(FILL_ME_IN, array.Length)
        Assert.Equal(FILL_ME_IN, array(0))
    End Sub

    <Koan(7)> _
    Public Sub TwoDimensionalArrays()
        Dim array(2, 3) As Integer
        array(2, 3) = 5
        Assert.Equal(FILL_ME_IN, array(2, 3))
    End Sub
End Class

File: Koans/CSharp/AboutArrays.cs
using System;
using System.Collections;
using System.Collections.Generic;
using Xunit;
using System.Linq;

namespace DotNetKoans.CSharp
{
    public class AboutArrays : Koan
    {
        [Koan(1)]
        public void CreatingAnArray()
        {
            var array = new int[3];
            Assert.Equal(typeof(int[]), array.GetType());
            Assert.Equal(4, array.Length);
        }

        [Koan(2)]
        public void ArrayLiterals()
        {
            // You can create an array and fill it with elements in one go,
            // in that case you don't need to specify the length.
            int[] array1 = new int[] { 42, 50 };
            Assert.Equal(typeof(int), array1[0].GetType());

            //You don't have to specify a type if the arguments can be inferred
            var array2 = new [] { 42, 50 };
            Assert.Equal(typeof(int), array2[0].GetType());

            // or even simpler:
            int[] array3 = { 42, 50 };
            Assert.Equal(typeof(int), array3[0].GetType());
        }

        [Koan(3)]
        public void AccessingElements()
        {
            int[] array1 = { 42, 50 };

            //Are arrays 0-based or 1-based?
            Assert.Equal(42, array1[((int)FILL_ME_IN)]);

            //This is important because...
            Assert.Equal(array1.IsFixedSize, FILL_ME_IN);

            //...it means we can't do this: array1[2] = 13;
            Assert.Throws(typeof(FillMeIn), delegate() { array1[2] = 13; });

            //This is because the array is fixed at length 1. You could write a function
            //which created a new array bigger than the last, copied the elements over, and
            //returned the new array. Or you could use a container, which we'll see later.
        }

        [Koan(4)]
        public void AccessingArrayElements()
        {
            var array = new[] { "peanut", "butter", "and", "jelly" };

            Assert.Equal(FILL_ME_IN, array[0]);
            Assert.Equal(FILL_ME_IN, array[3]);
            
            //This doesn't work: Assert.Equal(FILL_ME_IN, array[-1]);
        }

        [Koan(5)]
        public void SlicingArrays()
        {
            var array = new[] { "peanut", "butter", "and", "jelly" };

			Assert.Equal(new string[] { (string)FILL_ME_IN, (string)FILL_ME_IN }, array.Take(2).ToArray());
			Assert.Equal(new string[] { (string)FILL_ME_IN, (string)FILL_ME_IN }, array.Skip(1).Take(2).ToArray());
        }

        [Koan(6)]
        public void CreatingATwoDimensionalArray()
        {
            int[,] array = new int[3,2];
            array[2, 1] = 5;
            Assert.Equal(FILL_ME_IN, array[2, 1]);
        }

        [Koan(7)]
        public void CreatingATwoDimensionalArrayLiteral()
        {
            int[,] array = new int[3, 2]
            {
                { 1, 2 },
                { 3, 4 },
                { 5, 6 }
            };
            Assert.Equal(FILL_ME_IN, array[2, 1]);
        }

        [Koan(8)]
        public void CreatingAnArrayOfArrays()
        {
            // If you need a non-rectangular two dimensional array,
            // you need to create an array of arrays:
            int[][] array = new int[3][];
            array[0] = new int[1];
            array[1] = new int[2];
            array[2] = new int[3];

            array[2][2] = 5;
            Assert.Equal(FILL_ME_IN, array[2][2]);
        }
    }
}
